### PR TITLE
[transaction] Recover from failures during endTransaction after writing to storage

### PR DESCRIPTION
**Pull request by gaoran10**
(cherry picked from commit 7f59a829bfdfb5cffd7f4eda0dc8d549048c9e59) refer to https://github.com/datastax/starlight-for-kafka/commit/7f59a829bfdfb5cffd7f4eda0dc8d549048c9e59

### Motivation

Recover from failures during endTransaction after writing to storage.

### Modifications

Reset the transaction pending state from failures during endTransaction.  

### Documentation

Check the box below.

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [x] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionCoordinator.java [hunk 1]
```diff
@@ -613,6 +627,10 @@ public void fail(Errors errors) {
                             if (epochAndMetadata.getCoordinatorEpoch() == coordinatorEpoch) {
                                     // This was attempted epoch fence that failed, so mark this state on the metadata
                                 epochAndMetadata.getTransactionMetadata().setHasFailedEpochFence(true);
+
+                                // this line is not present in Kafka code base ?
+                                epochAndMetadata.getTransactionMetadata().setPendingState(Optional.empty());
```
Comment (gaoran10) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionCoordinator.java:632]: This is the key point.

Comment (Demogorgon314) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionCoordinator.java:632]: Okay, but why do we need to set the pending state to empty?

Comment (gaoran10) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionCoordinator.java:632]: If don't reset the pending state, it seems that the transaction coordinator will throw a concurrent exception when re-end the transaction, refer to [here](https://github.com/streamnative/kop/blob/master/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionCoordinator.java#L669), but there is no test to cover this case yet.

**Comment by Demogorgon314** [comment]
It looks like the test failed. Can you check the failed test?

**Review by Demogorgon314 — APPROVED** [review]

**Pull request merged**